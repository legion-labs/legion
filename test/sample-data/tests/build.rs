use std::{
    env,
    path::{Path, PathBuf},
};

use legion_data_compiler::Manifest;

pub fn target_dir() -> PathBuf {
    env::current_exe()
        .ok()
        .map(|mut path| {
            path.pop();
            if path.ends_with("deps") {
                path.pop();
            }
            path
        })
        .unwrap_or_else(|| panic!("cannot find test directory"))
}

pub fn data_build_exe() -> PathBuf {
    target_dir().join(format!("data-build{}", env::consts::EXE_SUFFIX))
}

fn exec_create_build_index(
    path: impl AsRef<Path>,
    project: impl AsRef<Path>,
) -> std::io::Result<std::process::Output> {
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("create");
    command.arg(path.as_ref().to_str().unwrap());
    command.arg(format!("--project={}", project.as_ref().to_str().unwrap()));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

fn exec_data_compile(
    compile_path: &str,
    buildindex_path: &Path,
    destination: &Path,
) -> std::io::Result<std::process::Output> {
    let target = "game";
    let platform = "windows";
    let locale = "en";
    let mut command = std::process::Command::new(data_build_exe());
    command.arg("compile");
    command.arg(compile_path);
    command.arg(format!("--cas={}", destination.to_str().unwrap()));
    command.arg(format!("--target={}", target));
    command.arg(format!("--platform={}", platform));
    command.arg(format!("--locale={}", locale));
    command.arg(format!(
        "--buildindex={}",
        buildindex_path.to_str().unwrap()
    ));
    let output = command.output()?;
    assert!(output.status.success());
    Ok(output)
}

/// read the build index and remove non-deterministic data.
fn read_build_index(buildindex_path: &Path) -> String {
    let content = std::fs::read_to_string(&buildindex_path).expect("file content");
    let mut content = json::parse(&content).expect("valid json");
    content.remove("project_index");
    content.pretty(2)
}

#[test]
fn incremental_build() {
    let work_dir = tempfile::tempdir().unwrap();
    let sampledata_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let project = sampledata_dir.join("project.index");
    let buildindex = work_dir.as_ref().join("build.index");

    // create build index and do a source pull
    exec_create_build_index(&buildindex, project).expect("new build index");

    insta::assert_snapshot!("initial_index", read_build_index(&buildindex));

    // default root object in sample data
    // /world/sample_1.ent (offline_entity) => runtime_entity
    let root_entity = "d004cd1c00000000fcd3242ec9691beb|019c8223";

    //
    // first data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let manifest: Manifest = serde_json::from_slice(&out.stdout).expect("valid manifest");
    insta::assert_json_snapshot!("first_manifest", manifest);

    let first_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("first_index", first_buildindex);

    //
    // incremental data build
    //
    let out =
        exec_data_compile(root_entity, &buildindex, work_dir.path()).expect("build completed");

    let incremental_manifest: Manifest =
        serde_json::from_slice(&out.stdout).expect("valid manifest");

    insta::assert_json_snapshot!("incremental_manifest", incremental_manifest);

    let incremental_buildindex = read_build_index(&buildindex);
    insta::assert_snapshot!("incremental_index", incremental_buildindex);

    // incremental build with no changes should not affect the build index
    assert_eq!(first_buildindex, incremental_buildindex);
}
