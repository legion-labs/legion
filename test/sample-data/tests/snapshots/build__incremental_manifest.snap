---
source: test/sample-data/tests/build.rs
expression: incremental_manifest

---
{
  "compiled_resources": [
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee",
      "checksum": "2836be573f8af76ad1ad8a3b77db56dd",
      "size": 2719561
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_Black Background",
      "checksum": "9fe86657cd4128af3319d454523b83da",
      "size": 2500086
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_Ferris",
      "checksum": "cb819d63f19c1a7a180c5973b9099232",
      "size": 2500672
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_albedo",
      "checksum": "2ca82c862c75b07586972883955dfc1e",
      "size": 2500672
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_metalness",
      "checksum": "2ca82c862c75b07586972883955dfc1e",
      "size": 2500672
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_normal",
      "checksum": "2ca82c862c75b07586972883955dfc1e",
      "size": 2500672
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_roughness",
      "checksum": "2ca82c862c75b07586972883955dfc1e",
      "size": 2500672
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_roughness|0659c0b6c23b950d",
      "checksum": "4d74c6b043ec45baabdf4a838d206c2e",
      "size": 1000038
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_normal|0659c0b6c23b950d",
      "checksum": "4d74c6b043ec45baabdf4a838d206c2e",
      "size": 1000038
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_metalness|0659c0b6c23b950d",
      "checksum": "4d74c6b043ec45baabdf4a838d206c2e",
      "size": 1000038
    },
    {
      "path": "(4976b194d8a1e0c2,8ee3dcef-1aa4-4772-ad9b-28066f06909e)|648242a8cf00bfee_albedo|0659c0b6c23b950d",
      "checksum": "4d74c6b043ec45baabdf4a838d206c2e",
      "size": 1000038
    },
    {
      "path": "(07dd9f5d1793ed64,48909c46-ad4f-4d6b-a522-2e16e81ba082)|738a5b850f6f8b3a",
      "checksum": "858de9bef4379b02fa58e90f40a47a1d",
      "size": 230
    },
    {
      "path": "(1c0ff9e497b0740f,270c28b5-6368-4a21-8ef1-a7d03f85abb3)|1d9ddd99aad89045",
      "checksum": "84c6349661f5a88ac385d3bafd2f0574",
      "size": 188
    },
    {
      "path": "(91dc60590cba171e,c6db12c5-47a6-407d-94da-b6853f36eb8a)|e5c099cc30495fed",
      "checksum": "f26271049c496a18c5a63e90676ff51b",
      "size": 87
    },
    {
      "path": "(91dc60590cba171e,7d0e930c-eb69-45c4-a695-fcb787f3700f)|e5c099cc30495fed",
      "checksum": "f26271049c496a18c5a63e90676ff51b",
      "size": 87
    },
    {
      "path": "(1c0ff9e497b0740f,45d44b64-b97c-486d-a0d7-151974c28263)|1d9ddd99aad89045",
      "checksum": "15b73f1772ab679e76702d8342504dd5",
      "size": 235
    },
    {
      "path": "(80b97dd7e2e9edca,7948dc8c-8bc6-4aaf-b3cf-0a21633edc44)|9c9a840b84935d1f",
      "checksum": "b3727675aedad2e3a72ec704485fef6d",
      "size": 967
    },
    {
      "path": "(91dc60590cba171e,ee30f104-3ef5-46e7-af16-4f085bd6f8e3)|e5c099cc30495fed",
      "checksum": "f26271049c496a18c5a63e90676ff51b",
      "size": 87
    },
    {
      "path": "(91dc60590cba171e,7d083d64-cbf9-4786-ae8b-44924b3e2035)|e5c099cc30495fed",
      "checksum": "f26271049c496a18c5a63e90676ff51b",
      "size": 87
    },
    {
      "path": "(91dc60590cba171e,4f595c8b-3b9a-41d6-9cf6-a7557dd2e8ee)|e5c099cc30495fed",
      "checksum": "f26271049c496a18c5a63e90676ff51b",
      "size": 87
    },
    {
      "path": "(91dc60590cba171e,181f9358-000f-47db-9583-7c9448c3dba8)|e5c099cc30495fed",
      "checksum": "03737357aaac789861d40c51f8cf3032",
      "size": 87
    },
    {
      "path": "(1c0ff9e497b0740f,e07c7027-a823-47e5-b0dd-8871e27adfce)|1d9ddd99aad89045",
      "checksum": "701b20b1a8aaabf3e08c84fae9df658c",
      "size": 286
    },
    {
      "path": "(09cde0390abe0694,db051b98-6ff5-4bac-bea8-50b5a13c3f1b)|e0d46a6b35b5717e",
      "checksum": "2ab41497efd7e6e832c9c108143e02a0",
      "size": 102
    },
    {
      "path": "(09cde0390abe0694,7483c534-fe2a-4f16-b655-e9afe39a93ba)|e0d46a6b35b5717e",
      "checksum": "ba97d1c2f46ebbae9f32edcb1e8f4972",
      "size": 102
    },
    {
      "path": "(09cde0390abe0694,202e3aa6-f158-4c77-890b-3f59b183b6bd)|e0d46a6b35b5717e",
      "checksum": "82a302e1cad5a0a885a56173419c85cf",
      "size": 102
    },
    {
      "path": "(1c0ff9e497b0740f,5d6c8521-ef7f-4402-8fac-01c5c4f53329)|1d9ddd99aad89045",
      "checksum": "d9d62a8fb32f5279575325d40f1bf3b8",
      "size": 586
    }
  ]
}
